### DISCUSSION TITLE
Ensoniq Mirage "ROMpler"

### DISCUSSION THREAD

**Discussion by roguestreak** [marcel-licence/esp32_midi_sampler#90] [Ideas]
Hi Marcel
Following the "Grand Piano" discussion in the basic_synth section, I've been off investigating the disk image format of the ancient (~1984) Ensoniq Mirage sampler, to get some info together before putting forward a feature/project suggestion.  I've gotten the major basics together, so before I go any deeper I'm seeing how feasible/useful the following is, what further info you'd need, etc.

The Ensoniq Mirage 8bit sampler had a massive(!) 64kb of memory for samples, but was used successfully on records for the likes of Janet Jackson, Skinny Puppy, Public Enemy, The Cure, etc., and is still held in high esteem by some.  There are literally hundreds of professionally-sampled disk images floating around the internet, making it potentially a great candidate for an ESP32 / RP2040 "rompler", ie. sample player.  Additionally it had 8x analogue VCFs and VCAs, so was more a sampling synthesizer than a straight sampler.

(it actually has TWO sets of 64kb memory, a "lower sound" and "upper sound" bank, but each loads independently, so can be treated as a separate entity if required for a new project.)

The available disk images, in EDM or IMG format (same structure, different name), are actually quite straight-forward to extract sample and parameter data from once you know the structure; you can import one in Audacity as "Raw Data -> unsigned 8bit, mono" and see the waveforms blatantly sitting there, but with "strips" of parameter and OS data interspersed.

So the question is, is it possible/feasible to be able to select an image from an SD card and load/extract either a single "bank", or both an upper and lower bank?
Image files are 440kb (450,560 bytes) and are uncompressed, so to my untrained mind it should be possible to read in a single bank directly from the file?
Each image ("disk") contains 3x lower banks and 3x upper banks; on a physical Mirage you would either "Load Lower: [1-3]", "Load Upper: [1-3]", or "Load Lower & Upper (Load All): [1-3])
Accompanying each bank are 8 sets of parameters, as each sound bank can be split into up to 8 different samples; these can be used as individual voices (single oscillator), or layered as two samples per voice with static or dynamic mixing (2 oscillators with different waveforms).  Parameter data includes loop points, sample start/end locations, key assignments, etc.  Sample data would be 65,536 bytes, and parameter data 1024 bytes.

I currently have figured out the data structure, where to start/stop reading for a bank and how to strip out the extraneous data, where parameters are located and what most of them are, and where general keyboard configuration parameters are stored; I can continue to track down exactly where every required parameter is located, but figured I'd see if the idea is technically feasible before mapping out every single piece.

If feasible, do you think a board with PSRAM would be a minimum requirement, or could it all happen with internal memory?

Some examples of "stock" sample disks in action:
https://www.youtube.com/watch?v=yPtTEzR3m-I

Sincerely, Lee

**Comment by roguestreak** [comment]
An example collection for download can be found here:
https://github.com/ZAPPPP/Mirage

IMG files are available individually to check out, the Factory Library for example has semi-descriptive disk labels so you can make an informed choice; "B2" is "piano" on banks #1, "electric piano" on banks #2, and "drawbar organ" on banks #3 
https://github.com/ZAPPPP/Mirage/tree/main/FactorySoundLibrary

**Reply by marcel-licence** [reply]
That is nice. Do you know how to extract the waveforms from the binary .img-files?

**Comment by roguestreak** [comment]
Hi Marcel
I don't know what the best / most efficient way of doing it is, but essentially from a certain starting point you read 5120 bytes, skip 512 bytes, and repeat another 12x times.  The first 1024 bytes read is sound parameters, the rest is sound data; presumably you'd want to split those into two different arrays?

Refer to this old diagram to help come to grips with the following info, as it's a bit confusing without a visual reference
http://www.synthzone.com/midi/ensoniq/mirage/The%20Mirage%20Monitor%20Disk_files/sectmap.gif
(I've made up an excel spreadsheet which goes into more detail, but it's on my work computer so no access today)

Disks were originally a custom-formatted Single-Sided Double Density 3.5" floppy
A Disk (image) has 80 "tracks", numbered 0 - 79
each consisting of 5x "Large" 1024 byte sectors,  plus 1x "Small" 512 byte sector
for a total of 5632 bytes per track
and a disk image total of 450,560 bytes

For the most part we can completely ignore the "Small" sectors, so they need to be skipped over / stripped out.

An example read:
LowerSound1
- start 11264  (track 2, 2x 5632)
  - read 5x 1024, skip 512
  - repeat 12x more times (13 total)

This gives 65,560 bytes; the first 1024 is the parameters, then 65,536 is the actual sound data for that "bank".


Here's the other Sound Bank starting points:

UpperSound1
- start 84480 (track 15, 15x 5632)

LowerSound2
- start 157696 (track 28, 28x 5632)

UpperSound2
- start 230912 (track 41, 41x 5632)

LowerSound3
- start 304128 (track 54, 54x 5632)

UpperSound3
- start 377344 (track 67, 67x 5632)


******************************************************************

I'm still figuring out exactly what the required parameters are and where they live, and also what comes under "configuration" rather than parameters (see final line).  Further confusing is that there can be 8x sets of parameters for the waveforms in a bank (eg. start/stop locations, loop points), plus 4x sets of Bank parameters for the like of the VCF and VCA envelopes.  A "parameter sheet" can be found here for reference:
http://studiorepair.com/gallery/Ensoniq/Mirage/slides/Ensoniq_Mirage_Parameter_Reference.html

But I have tracked down quite a few; I will continue mapping out if you think it's a feasible way of getting good samples in?

sound params for Lower Bank #1
start 11,264, viewed as three rows of 8 bytes (24 bytes total) in a binary editor; we'll call this Set #1, there are 8x sets of 24bytes which define the whole set of samples in the Lower Bank

d0 b3 de b3 e2 b3 e0 b3
e2 01 04 17 3f 2e c6 09
a0 bf be bf 80 00 00 00

- haven't correlated out the first row yet; I'm attempting to look at certain disk images after booting them up in my Mirage and writing down parameters I'm interested in, and trying to find them / patterns in the data

- 2nd row:
 3rd byte (04) = param 67 (relative tuning)
 4th byte (17) = param 68 (tuning fine)
 8th byte (09) = param 72 (top key, minus 1, converted to hex) -> top key in decimal for this sample is "10"

- third row (starting at byte 11,280 of the disk img)
 first five bytes are params 60, 61, 62 63, 64
 60 = wavesample start
 61 = wavesample end
 62 = loop start point
 63 = loop end point
 64 = loop end fine adjustment

Note that the Mirage's interface is usually in Hex (!!).  As such, samples are referenced by 256 "rough" groups of 256 "fine" bytes.  In the above loop parameters, only the "loop end fine adjustment" goes into byte-level, the others are referring to "pages" of 256 bytes

eg. a "wavesample start" of "6B" would equate to the start of the 107th page of 256bytes in memory, = byte 27,392


wavesample 2 param starts 11,288
wavesample 3 param starts 11,312
wavesample 4 param starts 11,336
etc etc., incrementing by 24


A note about "Top Key"; the wavesamples layer onto the keyboard
eg. first with top key 10 covers keys 0-10
next with top key 19 then covers 11-19
next with top key 27 then covers 20-27
etc

There's probably a whole lot of detail that needs to be covered off to get fully operational, but the actual sample data and start / end / loop points should be relatively straightforward


****************************************************************

Configuration is found in Small sector (512byte) of track 11, starting 67072

**Comment by roguestreak** [comment]
PS. You probably have a hex editor already and know what you're looking for, but for those playing along at home I'm using "Free Hex Editor Neo" to try to find the parameters, setting the view to "8 columns" and the left-hand view to "decimal".  Note for example here the "09" (which is the Hex value for param 72 "top key") is repeated three lines later as it's also used for sample 2, then is replaced by "12" for the next two sets, then "1a", then "23" when converted to Binary and one added (keys in memory start at "0") they come out as Top Keys of 10 (hex 09), 19 (hex 12), 27 (hex 1a) and 36 (hex 23)

![image](https://user-images.githubusercontent.com/34874261/212573547-244cf411-ca5e-4a8b-9751-37915f80d12c.png)

The Mirage interface was a mix of Hex and Binary, some things were listed in Binary (most "human interface" settings like envelopes etc) and some (mostly to do with wavesamples) were in Hex.  I've since found some text-file versions of the parameter cards, so listing them below.  Some aren't required/relevant, but leaving them in for completeness.

General Keyboard
----------------
21	Master Tune		00-99
22	Pitch Bend Range	00-34
23	Keyboard Vel. Sens.	00-63
24	Keyboard Balance	00-63
25	U/L Program Link	On-Off

WaveSample
----------
26	Wavesample Select	01-08
27	Initial Wavesample	01-08
28	Mix Mode		On-Off
33	Osc. 2 Detune		00-99
34	Osc. Mix		00-63
35	Osc. Mix-Vel. Sens.	00-31, 00=Mod. Wheel

Keyboard/Program
----------------
29	Monophonic Mode		On-Off
31	LFO Freq. (Speed)	00-99
32	LFO Depth		00-99, 00=Mod. Wheel
36	Filter Cutoff Freq.	00-99
37	Filter Resonance (Q)	00-40
38	Filter Kbd. Tracking	00-04

Amplitude Envelope
------------------
50	Attack			00-31
51	Peak			00-31
52	Decay			00-31
53	Sustain			00-31
54	Release			00-31

Amplitude Envelope Modulation
-----------------------------
55	Attack-Vel. Sens.	00-31
56	Peak-Vel. Sens.		00-31
57	Decay-Kbd. Scaled	00-31
58	Sustain-Vel. Sens.	00-31
59	Release-Vel. Sens.	00-31

Filter Envelope
---------------
40	Attack			00-31
41	Peak			00-31
42	Decay			00-31
43	Sustain			00-31
44	Release			00-31

Filter Envelope Modulation
--------------------------
45	Attack-Vel. Sens.	00-31
46	Peak-Vel. Sens.		00-31
47	Decay-Kbd. Scaled	00-31
48	Sustain-Vel. Sens.	00-31
49	Release-Vel. Sens.	00-31

Sampling Configuration
----------------------
73	Sample Time Adj.	30-99
74	Input Filter Freq.	00-99
75	Line/Mic Level Input	On-Off
76	Sampling Threshold	00-63
77	User Multisampling	On-Off

Command
-------
11	Save Lower Kbd. Sound
12	Save Upper Kbd. Sound
13	Save Both Kbd. Sounds
14	Save Config. Params
15	Copy Program to Lower
16	Copy Program to Upper

MIDI Configuration
------------------
81	MIDI Omni Mode		On-Off
82	MIDI Channel		01-16
83	MIDI Thru Mode		On-Off
84	MIDI Controller Enable	On-Off
85	Ext. Sequener Clock	On-Off
86	Ext. Clock Jack Select	On-Off
87	Internal Clock Rate	00-99
88	Sequencer Loop Switch	On-Off
89	Seq. Ft. Sw./Sus. Pedal	On-Off

Wavesample
----------
26	Wavesample Select	01-08
27	Initial Wavesample	01-08
60	Wavesample Start	00-ff
61	Wavesample End		00-ff
62	Loop Start		00-ff
63	Loop End		00-ff
64	Loop End Fine Adj.	00-ff
65	Loop Switch		On-Off
66	Wavesample Rotate	00-ff
67	Relative Tuning-Coarse	00-07
68	Relative Tuning-Fine	00-ff
69	Relative Amplitude	00-63
70	Relative Filter Freq.	00-99
71	Maximum Filter Freq.	00-99
72	Top Key			01-61

**Comment by roguestreak** [comment]
Hi Marcel
I believe I've tracked down all of the required parameters and how they are stored; there are some oddities, eg. some params are stored as double their front panel value, some quadruple, but it turns out this is all noted in the Mirage Advanced Sampler's Guide, I just didn't know how to interpret it previously.  Should again be pretty straight-forward to extract now that I know where it is.
It does however raise the complexities of the Mirage's envelope and mixing system, which it's probably best to read through said Guide to see how it might impact any implementation, whether certain elements should just be discarded, hard-preset etc rather than following the dynamicism of the Mirage's setup?
One point is case is the filter envelopes, which use APDSR, with the extra Peak and the Sustain both being capable of varying with velocity, and thus affecting the rest of the envelope, plus keyboard tracking.  And each sample has its own "relative" settings which are added to the "base" settings; this is designed so different samples in a Bank can be turned up or down, made brighter etc., to have them fit in with the other samples better.
![image](https://user-images.githubusercontent.com/34874261/213847850-e7e40b2b-51af-463a-a14d-9e4656e59798.png)

It just takes a bit of study to comprehend, perhaps then certain elements can be simplified, eliminated, or incorporated?
https://github.com/ZAPPPP/Mirage/blob/main/Manuals/ensoniq_mirage_advanced_sampler's_guide.pdf

PS. the four Parameters per Bank can be seen as pre-sets; not only can they change the envelopes and mixes, but they can also change the sample selection within the bank itself.  So for example, Bank #1 of the "B2" disk linked above has the capability, using the 4 presets,  of doing two different acoustic pianos, a "honky-tonk" piano, and something that may perhaps be a koto?  All with just 64kb of samples for the "lower half" of the keyboard

**Comment by roguestreak** [comment]
Or Program (Preset) #1 could just be loaded and ignore the others, at least for initial testing?

**Comment by roguestreak** [comment]
PPS. The actual synth structure is stock-standard 80s though, it's just the way the mixes / envelopes are done can by impacted by multiple factors
![image](https://user-images.githubusercontent.com/34874261/213859062-9e81c74e-25fd-4b3b-94cf-c2d03088588e.png)

**Reply by cbrightly** [reply]
It’s 2026 and an Ensoniq-ish 8/12 bit sampling drum machine based on ESP32? The project sound amazing … status update?

**Comment by gordonjcp** [comment]
https://github.com/gordonjcp/miragetools

Partial disassembly of the ROM and OS3.2 disk. I know a bit about how the sounds and stuff are laid out, if you want some help.

**Comment by roguestreak** [comment]
Sorry I haven't been checking my mail (secondary account), no updates on my front, I never even "got around to" testing out one of Marcel's standard offerings first, always had that one thing that was stopping me, mostly making the MIDI interface.
Gordon I think I got the majority of my starting points from info provided by you, then a few blanks had been filled by just experimenting, eg. finding a disk with know parameter values and seeing if I could correlate those values to locations in the file.
Ironically my Mirage has had a bit of a resurgence of late thanks to a modified gotek drive, but I still do believe the disk images and structure could be a great basis for a professionally sounding rompler using minimalistic cheap hardware.  Happy to progress further down this route, but DSP is not my thing and I haven't had any luck at all deciphering how Marcel works his magic.

PS. Looks like the SynthZone image linked above is now dead, but the wayback machine can assist
https://web.archive.org/web/20241227040729/http://www.synthzone.com/midi/ensoniq/mirage/The%20Mirage%20Monitor%20Disk_files/sectmap.gif
